PHIElimination: add target hook to control reuse.

jgu222 opened the pull request:
PHI elimination pass may reuse temporary registers during PHI node elimination if two phis have the same right-hand sites.

However, the left-hand sides of phis need to be checked as well for some GPU targets,, such as an Intel GPU target. If the register classes of phis's left-hand sides are different, reuse might be problematic, as shown in https://github.com/llvm/llvm-project/issues/163500. Thus, this change adds additional check for reuse. 

The additional check is to check if the left-hand sites of phis have the same common sub register class. If so, reuse is allowed. Otherwise, resuse is not allowed. This might be conservative, but expect no impact to CPU code.

This is to resolve the issue:
  https://github.com/llvm/llvm-project/issues/163500

jgu222 committed:
PHIElimination: add target hook to control reuse.

Add PHI reuse hook so that a target can decide if temporary register
during PHI node elimination can be reused. By default, two phis are
allowed to use the same temporary register if their right-hand-sides
are the same.

However, the left-hand sides of phis need to be checked as well
for some targets, such as a GPU target. If the register banks of
phis's left-hand sides are diffrerent, reuse is not allowed.
Thus, this change adds a target hook for this kind of checking
as it is target-dependent.

This change has no functional change.

This is to resolve the issue:
  https://github.com/llvm/llvm-project/issues/163500

jgu222 commented:
@paperchalice @guy-david   Could you help review this ?

jgu222 commented:
@arsenm Could you help review it ?

jgu222 commented:
`This sounds more like a modeling issue rather than a problem with this specific pass.`
Not sure about modeling. But if so, what would be a solution ? Could you elaborate on that ?

To me, the phi node elimination assumes that two phis that have identical right-hand sides can use the same virtual temp. This is true for cpu, but not true for gpu due to gpu code's complicate divergence/convergence scheme. 

If not this one, what approach is more appropriate ?  Thanks

arsenm reviewed (changes requested):
This is too specific and has no test coverage. There are existing methods for checking if registers share a register bank or class

jgu222 commented:
> This is too specific and has no test coverage. There are existing methods for checking if registers share a register bank or class

Kind of agree that this is too specific.  Using register class works for me,  although I feel this checking  is a stronger condition.  But I can live with it.

How is the following change ?
```
diff --git a/llvm/lib/CodeGen/PHIElimination.cpp b/llvm/lib/CodeGen/PHIElimination.cpp
index 34a9d5d0e401..5f08aa120414 100644
--- a/llvm/lib/CodeGen/PHIElimination.cpp
+++ b/llvm/lib/CodeGen/PHIElimination.cpp
@@ -377,10 +377,17 @@ void PHIEliminationImpl::LowerPHINode(MachineBasicBlock &MBB,
     // typically those created by tail duplication. Typically, an identical PHI
     // node can't occur, so avoid hashing/storing such PHIs, which is somewhat
     // expensive.
+    auto canReuse = [&MF](Register Reg0, Register Reg1) {
+      auto &MRI = MF.getRegInfo();
+      auto *RC0 = MRI.getRegClass(Reg0);
+      auto *RC1 = MRI.getRegClass(Reg1);
+      return RC0 && RC1 && RC0 == RC1;
+    };
+
     Register *Entry = nullptr;
     if (AllEdgesCritical)
       Entry = &LoweredPHIs[MPhi];
-    if (Entry && *Entry) {
+    if (Entry && *Entry && canReuse(*Entry, DestReg)) {
       // An identical PHI node was already lowered. Reuse the incoming register.
       IncomingReg = *Entry;
       reusedIncoming = true;

```

arsenm commented:
Is this the problem solved by SIOptimizeVGPRLiveRange?

You generally want to work in terms of common subclasses than exact class checks

jgu222 committed:
Use common register class

Remove target hook. Instead, use common register class to
check if two phis are allowed to reuse the same temp.

jgu222 commented:
> I don't think this is the correct approach, but correct me if I'm mistaken. The underlying assumption of reusing the same virtual destination register is that in the worst case, values will be copied around (different register classes/banks, allocation). In the presented issue, it seems like there's an actual semantic difference between what the virtual registers represent in each block, depending on which thread is executing? This sounds more like a modeling issue rather than a problem with this specific pass. Maybe verifying that the register classes overlap can be a workaround in this specific case, but it seems like more issues will pop along the way.

PR has been re-done.  Could you comment ?  If okay,  can you help approve and merge ?

jgu222 commented:
I feel there is misunderstanding here. Here is what I think this pass should do:
  1.  PHIElimination is generic and its translation is expected to be functionally correct for all targets
  2.  reuse in PHIElimination is an optional optimization. PHIElimination should function correctly with or without this reuse.
  
The reason for this PR is that (1) does not hold for our target.  We can surely work-around it.  But if so, why do we always perform this optimization when we know it is not correct for all targets ?  Especially,  this optimization is optional and potentially expensive ?

jgu222 commented on llvm/test/CodeGen/AMDGPU/phi-elimination-reuse-pr163604.mir:
Our target (internal, non public) allows this. I use this just to show what this PR is doing. Looks like this test should be removed.

arsenm replied:
What does "this" mean? You have different register classes, this is an unrepresentable situation

jgu222 replied:
It is scalar register <-> vector register copying (first active lane, implementation dependent).
Anyway,  I have to get around it.

jgu222 commented:
No longer need this. Close it
